### ISSUE TITLE
spork + cucumber-rails = "undefined method `World' for main:Object (NoMethodError)"

### ISSUE THREAD

**Issue by citrus** [sporkrb/spork#122]
**UPDATE:** Fix by upgrading to [spork-0.9.0-rc9](https://rubygems.org/gems/spork/versions/0.9.0.rc9)

---

Hello,

I'm trying to use Spork with Rails and Cucumber via cucumber-rails and keep running into a NoMethodError. 

Here's the steps I took to create this error:

```
rvm use ruby-1.9.2-p180@sporkfail --create
gem install bundler
gem install rails --pre
rails new sporkfail
cd sporkfail

# edit Gemfile

bundle install
rails generate cucumber:install --spork
rails generate cucumber:feature post title:string body:text published:boolean
rails generate scaffold post title:string body:text published:boolean
rake db:migrate

bundle exec rake cucumber
# works

bundle exec spork
# error
```

Here's what's in the my Gemfile

``` ruby
source 'http://rubygems.org'

gem 'rails', '3.1.0.rc4'
gem 'sqlite3'

gem 'sass-rails', "~> 3.1.0.rc"
gem 'coffee-script'
gem 'uglifier'
gem 'jquery-rails'

group :test do
  gem 'cucumber-rails',    '>= 1.0.0'
  gem 'database_cleaner',  '>= 0.6.7'
  gem 'spork',             '>= 0.9.0.rc8'
end
```

And Here is the error I'm getting:

``` sh
$ bundle exec spork
Using Cucumber
Preloading Rails environment
Loading Spork.prefork block...
undefined method `World' for main:Object (NoMethodError)
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/cucumber-rails-1.0.0/lib/cucumber/rails/world.rb:24:in `<top (required)>'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:237:in `require'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:237:in `block in require'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:225:in `load_dependency'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:237:in `require'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/cucumber-rails-1.0.0/lib/cucumber/rails.rb:4:in `<top (required)>'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:237:in `require'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:237:in `block in require'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:223:in `block in load_dependency'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:639:in `new_constants_in'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:223:in `load_dependency'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:237:in `require'
/Users/Spencer/RoR/rails31/sporkfail/features/support/env.rb:11:in `block in <top (required)>'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/spork-0.9.0.rc8/lib/spork.rb:24:in `prefork'
/Users/Spencer/RoR/rails31/sporkfail/features/support/env.rb:10:in `<top (required)>'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:233:in `load'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:233:in `block in load'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:223:in `block in load_dependency'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:639:in `new_constants_in'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:223:in `load_dependency'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/activesupport-3.1.0.rc4/lib/active_support/dependencies.rb:233:in `load'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/spork-0.9.0.rc8/lib/spork/test_framework.rb:138:in `block (2 levels) in preload'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/spork-0.9.0.rc8/lib/spork/app_framework/rails.rb:8:in `preload'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/spork-0.9.0.rc8/lib/spork/test_framework.rb:134:in `block in preload'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/spork-0.9.0.rc8/lib/spork.rb:62:in `exec_prefork'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/spork-0.9.0.rc8/lib/spork/test_framework.rb:120:in `preload'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/spork-0.9.0.rc8/lib/spork/test_framework/cucumber.rb:17:in `preload'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/spork-0.9.0.rc8/lib/spork/run_strategy/forking.rb:25:in `preload'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/spork-0.9.0.rc8/lib/spork/runner.rb:74:in `run'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/spork-0.9.0.rc8/lib/spork/runner.rb:10:in `run'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/gems/spork-0.9.0.rc8/bin/spork:10:in `<top (required)>'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/bin/spork:19:in `load'
/Users/Spencer/.rvm/gems/ruby-1.9.2-p180@sporkfail/bin/spork:19:in `<main>'
```

Thanks!

-Spencer

**Comment by sporkd** [comment]
+1

Here are some other sightings...
https://github.com/cucumber/cucumber-rails/issues/148

**Comment by jkrall** [comment]
I ran into this tonight... and came up with a temporary fix, but I'm unsure what the proper resolution is or I'd put together a pull request.

in spork/test_framework/cucumber.rb, the following change fixes this issue:

```
def preload
  require 'cucumber'
  # if ::Cucumber::VERSION >= '0.9.0'
  #  # nothing to do nowadays
  # else
    preload_legacy_cucumbers
  # end
  super
end
```

[This commit](https://github.com/timcharper/spork/commit/cf9aeb12b34babd037812176fe3100565c85a190#diff-0) appears to be attempting to make a change to bring spork compatible with cucumber 0.9.0, but maybe something changed again in 1.0... because now it definitely seems like these commands in `preload_legacy_cucumbers` are required:

```
  @step_mother = ::Cucumber::Runtime.new
  @step_mother.load_programming_language('rb')
```

**Comment by jkrall** [comment]
Hmm... looks like the above fix doesn't really solve the problem.  

With the above fix in place, running cukes inside spork produces a plain Object inside step-definition blocks, instead of a Cucumber::Rails::World object which has all of the driver methods included.  This results in errors like:

```
NoMethodError: undefined method `visit' for #<Object:0x00000101f9e0a0>
./features/step_definitions/web_steps.rb:50:in `/^(?:|I )go to (.+)$/'
```

I tried downgrading cucumber & cucumber-rails to get around this problem... but unfortunately rails 3.1 requires builder-3.0.0, and cucumber ~> 0.9.0 requires builder ~> 2.1.2.

**Comment by larryh** [comment]
Got the same thing. Looks like this is definitely on the radar of the right people. Thanks for pointing this out... and thanks to those who provide the gems and fix them when something goes wrong!

**Comment by codebeige** [comment]
+1

**Comment by arunsark** [comment]
I am also facing the same issue

**Comment by jbarreneche** [comment]
@jkrall, did you also changed the #run_tests?

``` ruby
  def preload
    require 'cucumber'
    #if ::Cucumber::VERSION >= '0.9.0'
      # nothing to do nowadays
    #else
      preload_legacy_cucumbers
    #end
    super
  end

  def run_tests(argv, stderr, stdout)
    #if ::Cucumber::VERSION >= '0.9.0'
    #  ::Cucumber::Cli::Main.new(argv, stdout, stderr).execute!
    #else
      ::Cucumber::Cli::Main.new(argv, stdout, stderr).execute!(@step_mother)
    #end
  end
```

With both changes I managed to run my features with spork.

**Comment by citrus** [comment]
@jbarreneche - Your fix works for me as well. Thanks!!

**Comment by jkrall** [comment]
Nope, I didn't try that... good find!  Thanks for the tip.

**Comment by stepheneb** [comment]
also see: https://github.com/timcharper/spork/pull/127

(don't know how to attach a new pull request to an existing issue)

**Comment by stepheneb** [comment]
Tim merged my pull request into master: https://github.com/timcharper/spork/pull/127

I'd like to hear from others if this change works for them also?

**Comment by larryh** [comment]
It worked for me - thanks!

**Comment by gberenfield** [comment]
spork-0.9.0-rc9 fixed this for me. Thank you!!

**Comment by citrus** [comment]
Fixed with spork-0.9.0-rc9 - thanks @timcharper & @stepheneb

**Comment by timcharper** [comment]
thank you @stepheneb, all I did is check my email, look over his patch, and click merge. :)
